src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs

```diff
@@ -257,86 +258,377 @@ private struct IntPtrAndHandle
+            else if (type == typeof(SafeHandle))
+                return MarshalType.SAFEHANDLE;
+            else if (typeof(Delegate).IsAssignableFrom(type))
+                return MarshalType.DELEGATE;
+            else if ((type == typeof(Task)) || typeof(Task).IsAssignableFrom(type))
+                return MarshalType.TASK;
+            // HACK: You could theoretically inherit from Uri, but I consider this out of scope.
+            // If you really need to marshal a custom Uri, define a custom marshaler for it
+            else if (typeof(Uri) == type)
+                return MarshalType.URI;
+            else if (type.IsPointer)
+                return MarshalType.POINTER;
+
+            if (type.IsValueType)
+                return MarshalType.VT;
+            else
+                return MarshalType.OBJECT;
+        }
+
+        private static char? GetCallSignatureCharacterForMarshalType (MarshalType t, char? defaultValue) {
+            switch (t) {
+                case MarshalType.BOOL:
+                case MarshalType.INT:
+                case MarshalType.UINT32:
+                    return 'i';
+                case MarshalType.UINT64:
+                case MarshalType.INT64:
+                    return 'l';
+                case MarshalType.FP32:
+                    return 'f';
+                case MarshalType.FP64:
+                    return 'd';
+                case MarshalType.STRING:
+                    return 's';
+                case MarshalType.URI:
+                    return 'u';
+                case MarshalType.SAFEHANDLE:
+                    return 'h';
+                case MarshalType.ENUM:
+                    return 'j';
+                case MarshalType.ENUM64:
+                    return 'k';
+                case MarshalType.OBJECT:
+                    return 'o';
+                case MarshalType.VT:
+                    return 'a';
+                case MarshalType.POINTER:
+                    return 'm';
+                default:
+                    if (defaultValue.HasValue)
+                        return defaultValue;
+                    else
+                        throw new WasmInteropException($"Unsupported marshal type {t}");
             }
-            return new string(res);
+        }
+
+        public static string GetCallSignature(IntPtr methodHandle, object objForRuntimeType)
+        {
+            IntPtrAndHandle tmp = default(IntPtrAndHandle);
+            tmp.ptr = methodHandle;
+
+            MethodBase? mb = objForRuntimeType == null ? MethodBase.GetMethodFromHandle(tmp.handle) : MethodBase.GetMethodFromHandle(tmp.handle, Type.GetTypeHandle(objForRuntimeType));
+            if (mb == null)
+                return string.Empty;
+
+            ParameterInfo[] parms = mb.GetParameters();
+            int parmsLength = parms.Length;
+            if (parmsLength == 0)
+                return string.Empty;
+
+            var sb = new StringBuilder();
```

Comment:
this could be also a simple `var sb = new char[parmsLength]`, right?

Reply:
As currently written, yes. Call signatures can conditionally be longer than the number of parameters (for the trailing 'unmarshalled result' specifier) but at present this code will never emit it, so I can switch to a char array.